index.js

```diff
@@ -113,10 +113,57 @@ const ipv6 = {
 
 const ipv6Address = address(ipv6)
 
+const ip = {
+  preencode (state, string) {
+    const family = string.includes(':') ? 6 : 4
+    c.uint8.preencode(state, family)
+    if (family === 4) ipv4.preencode(state)
+    else ipv6.preencode(state)
+  },
+  encode (state, string) {
+    const family = string.includes(':') ? 6 : 4
+    c.uint8.encode(state, family)
+    if (family === 4) ipv4.encode(state, string)
+    else ipv6.encode(state, string)
+  },
+  decode (state) {
+    const family = c.uint8.decode(state)
+    if (family === 4) return ipv4.decode(state)
+    else return ipv6.decode(state)
+  }
+}
+
+const ipAddress = {
+  preencode (state, m) {
+    const family = m.host.includes(':') ? 6 : 4
+    c.uint8.preencode(state, family)
+    if (family === 4) ipv4.preencode(state)
+    else ipv6.preencode(state)
+    port.preencode(state, m.port)
```

Comment:
couldn't you just do `ip.preencode(...); port.preencode(...)` here ? same below etc